[FEA] Allow users to specify custom Dependency jars

amahussein opened the pull request:
Signed-off-by: Ahmed Hussein <ahussein@nvidia.com>

Fixes #1359

Add a new input argument that takes a path to a yaml file `--tools_config_file` The config file allows the users to define their own binaries that need to be added to the classpath of the tools jar cmd. This change is important because users can use the user-tools wrapper with their custom spark.

### Sample usage and specifications

- the dependenices can be `https:` or `file:///`
- assume that the spark binaries are for `spark-3.3.1-bin-hadoop3` which exists on local disk: `/local/path/to/spark-3.3.1-bin-hadoop3.tgz`
- We can define the config-file `/home/tools-conf.yaml` as follows:
    ```yaml
    info:
      api_version: '1.0'
    runtime:
      dependencies:
        - name: my-spark331
          uri: file:///local/path/to/spark-3.3.1-bin-hadoop3.tgz
          dependency_type:
            dep_type: archive
            # for tgz files, it is required to give the subfolder where the jars are located
            relative_path: jars/*
    ```
- now, we can run the python command as :

  ```
  spark_rapids profiling \
    --tools_config_file /home/tools-conf.yaml \
    --output_folder $LOCAL_OUTPUT_DIR \
    --eventlogs $LOCAL_GPU_EVENT_LOGS \
    --tools_jar $LOCAL_JAR \
    --verbose
  ```

- Assume that there are more jar files needed for the classpath, then the user can define an array of dependencies
  ```yaml
  info:
    api_version: '1.0'
  runtime:
    dependencies:
      - name: Hadoop AWS
        uri: https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-aws/3.3.4/hadoop-aws-3.3.4.jar
        verification:
          size: 962685
      - name: AWS Java SDK Bundled
        uri: https://repo1.maven.org/maven2/com/amazonaws/aws-java-sdk-bundle/1.12.262/aws-java-sdk-bundle-1.12.262.jar
        verification:
          file_hash:
            algorithm: sha1
            value: 02deec3a0ad83d13d032b1812421b23d7a961eea
      - name: spark-3.5.0
        uri: https:///archive.apache.org/dist/spark/spark-3.5.0/spark-3.5.0-bin-hadoop3.tgz
        verification:
          file_hash:
            algorithm: sha512
            value: 8883c67e0a138069e597f3e7d4edbbd5c3a565d50b28644aad02856a1ec1da7cb92b8f80454ca427118f69459ea326eaa073cf7b1a860c3b796f4b07c2101319
        dependency_type:
          dep_type: archive
          relative_path: jars/*
  ```

<details><summary>The specification of the file is as follows </summary>
<p>

```yaml
---
"$defs":
  DependencyType:
    description: Represents the dependency type for the jar cmd
    enum:
    - jar
    - archive
    - classpath
    title: DependencyType
    type: string
  DependencyVerification:
    description: The verification information of the dependency.
    properties:
      size:
        default: 0
        description: The size of the dependency file.
        examples:
        - 3265393
        title: Size
        type: integer
      file_hash:
        "$ref": "#/$defs/FileHashAlgorithm"
        default: 
        description: The hash function to verify the file
        examples:
        - algorithm: md5
          value: bc9bf7fedde0e700b974426fbd8d869c
    title: DependencyVerification
    type: object
  FileHashAlgorithm:
    description: |-
      Represents a file hash algorithm and its value. Used for verification against an
      existing file.
      ```py
      try:
          file_algo = FileHashAlgorithm(algorithm=HashAlgorithm.SHA256, value='...')
          file_algo.verify_file(CspPath('file://path/to/file'))
      except ValidationError as e:
          print(e)
      ```
    properties:
      algorithm:
        "$ref": "#/$defs/HashAlgorithm"
      value:
        title: Value
        type: string
    required:
    - algorithm
    - value
    title: FileHashAlgorithm
    type: object
  HashAlgorithm:
    description: Represents the supported hashing algorithms
    enum:
    - md5
    - sha1
    - sha256
    - sha512
    title: HashAlgorithm
    type: string
  RuntimeDependency:
    description: |-
      The runtime dependency required by the tools Jar cmd. All elements are downloaded and added
      to the classPath.
    properties:
      name:
        description: The name of the dependency
        title: Name
        type: string
      uri:
        anyOf:
        - format: uri
          minLength: 1
          type: string
        - format: file-path
          type: string
        description: The FileURI of the dependency. It can be a local file or a remote
          file
        examples:
        - file:///path/to/file.jar
        - https://mvn-url/24.08.1/rapids-4-spark-tools_2.12-24.08.1.jar
        - gs://bucket-name/path/to/file.jar
        title: Uri
      dependency_type:
        "$ref": "#/$defs/RuntimeDependencyType"
        description: The type of the dependency and how to find the lib files after
          decompression.
      verification:
        "$ref": "#/$defs/DependencyVerification"
        default: 
        examples:
        - size: 3265393
        - fileHash:
            algorithm: md5
            value: bc9bf7fedde0e700b974426fbd8d869c
    required:
    - name
    - uri
    title: RuntimeDependency
    type: object
  RuntimeDependencyType:
    description: |-
      Defines the type of dependency. It can be one of the following:
         - Archived file (.tgz)
         - Simple JAR file (*.jar)
         - Classpath directory (not yet supported)

      Note: The 'classpath' type is reserved for future use, allowing users to point to a directory
      in the classpath without needing to download or copy any binaries.
    properties:
      dep_type:
        "$ref": "#/$defs/DependencyType"
        description: The type of the dependency
      relative_path:
        default: 
        description: The relative path of the dependency in the classpath. This is
          relevant for tar files
        examples:
        - jars/*
        title: Relative Path
        type: string
    required:
    - dep_type
    title: RuntimeDependencyType
    type: object
  ToolsConfigInfo:
    description: |-
      High level metadata about the tools configurations (i.e., the api-version and any other relevant
      metadata).
    properties:
      api_version:
        description: The version of the API that the tools are using. This is used
          to test the compatibility of the configuration file against the current
          tools release
        examples:
        - '1.0'
        maximum: 1
        minimum: 1
        title: Api Version
        type: number
    required:
    - api_version
    title: ToolsConfigInfo
    type: object
  ToolsRuntimeConfig:
    description: The runtime configurations of the tools as defined by the user.
    properties:
      dependencies:
        description: The list of runtime dependencies required by the tools Jar cmd.
          All elements are downloaded and added to the classPath
        items:
          "$ref": "#/$defs/RuntimeDependency"
        title: Dependencies
        type: array
    required:
    - dependencies
    title: ToolsRuntimeConfig
    type: object
description: Main container for the user's defined tools configuration
properties:
  info:
    "$ref": "#/$defs/ToolsConfigInfo"
    description: Metadata information of the tools configuration
  runtime:
    "$ref": "#/$defs/ToolsRuntimeConfig"
    description: Configuration related to the runtime environment of the tools
required:
- info
- runtime
title: ToolsConfig
type: object

```
</p>
</details> 

### detailed code changes

This pull request includes several changes to the `user_tools` module, focusing on improving dependency management and configuration handling. The changes include importing new modules, adding methods for handling tool configurations, and updating the way dependencies are processed and verified.

#### Configuration Handling Enhancements:

* [`user_tools/src/spark_rapids_pytools/rapids/rapids_tool.py`](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0R75-R81): Added a method `get_tools_config_obj` to retrieve the tools configuration object from CLI arguments.
* [`user_tools/src/spark_rapids_pytools/rapids/rapids_tool.py`](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0R607-R619): Added a method `populate_dependency_list` to check for dependencies in a config file before falling back to default dependencies.

#### Minor Code Improvements:
* [`user_tools/src/spark_rapids_pytools/rapids/rapids_tool.py`](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0L139-R148): Corrected a comment typo in `_process_output_args` method.
* [`user_tools/src/spark_rapids_pytools/rapids/rapids_tool.py`](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0L396-R406): Adjusted the formatting of the `get_rapids_tools_dependencies` method for better readability.
* [`user_tools/src/spark_rapids_pytools/resources/dev/prepackage_mgr.py`](diffhunk://#diff-48f4c8d291d66fc725353b1a3d40b1a126a974b64c6e1daa149f3f765654fdc5L21): Added import for `CspPath` and `RuntimeDependency` to support new dependency handling. [[1]](diffhunk://#diff-48f4c8d291d66fc725353b1a3d40b1a126a974b64c6e1daa149f3f765654fdc5L21) [[2]](diffhunk://#diff-48f4c8d291d66fc725353b1a3d40b1a126a974b64c6e1daa149f3f765654fdc5L30-R30)

#### Modified the format of platform configuration to match the same structure

* [`user_tools/src/spark_rapids_pytools/rapids/rapids_tool.py`](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0L41-R45): Added imports for `RuntimeDependency`, `ToolsConfig`, and `DependencyType` to better manage dependencies. Updated methods to use `RuntimeDependency` objects instead of dictionaries. [[1]](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0L41-R45) [[2]](diffhunk://#diff-55690ae7ca276d65ab3f1ed28b6329ba3749b18ac3d5a6b9f22d5b81108390a0L535-R586)
* [`user_tools/src/spark_rapids_pytools/resources/databricks_aws-configs.json`](diffhunk://#diff-15ad57b52cf137d0237436d0e4658507fea041f8b955bbb1220f049052d1dd97L11-R27): Replaced `hashLib` with `fileHash` for verification and added `dependencyType` for better dependency classification. [[1]](diffhunk://#diff-15ad57b52cf137d0237436d0e4658507fea041f8b955bbb1220f049052d1dd97L11-R27) [[2]](diffhunk://#diff-15ad57b52cf137d0237436d0e4658507fea041f8b955bbb1220f049052d1dd97L36-R39) [[3]](diffhunk://#diff-15ad57b52cf137d0237436d0e4658507fea041f8b955bbb1220f049052d1dd97L50-R83)
* [`user_tools/src/spark_rapids_pytools/resources/databricks_azure-configs.json`](diffhunk://#diff-50b97b00c2b96b04b67a582b7826208b2c306260d5141b4ed434ce343bbcf29eL11-R59): Similar changes to the AWS configs, updating verification and dependency type fields.
* [`user_tools/src/spark_rapids_pytools/resources/dataproc-configs.json`](diffhunk://#diff-56742e03e6d5dc473bb7bcaecc167cd8af5651bed08e86c485129fab8e6631e9L11-R59): Updated verification and dependency type fields for Dataproc configurations.
* [`user_tools/src/spark_rapids_pytools/resources/dataproc_gke-configs.json`](diffhunk://#diff-27f1ca42cde0f902634144f1a166dd3d8f96f6c81bed5f94ff90ed737acac66cL11-R59): Similar updates to verification and dependency type fields for GKE configurations.

parthosa commented on user_tools/src/spark_rapids_pytools/resources/dev/prepackage_mgr.py:
```diff
@@ -100,34 +100,38 @@ def _fetch_resources(self) -> dict:
+                    uri_str = str(dependency.uri)
+                    pretty_name = FSUtil.get_resource_name(uri_str)
+                    resource_uris[uri_str] = {
+                        'depItem': dependency,
```
What would be the purpose of this key `depItem`? We seem to use `uri_str` and `prettyName` only.

amahussein replied:
Currently we do not.
I left it for future improvement and improvement in the script.
For example, the class can load all the configuration of the download entry from the json/yaml. Moving forward, if we have some changes in doing some verification/checks or extend the properties of the download item, then we won't have to do major changes in that script again.

parthosa reviewed (commented):
Thanks @amahussein for this PR. Adding a module for configurations using pydantic model was interesting. Minor comments.

amahussein commented on user_tools/src/spark_rapids_tools/configuration/common.py:
```diff
@@ -0,0 +1,87 @@
+                  'https://mvn-url/24.08.1/rapids-4-spark-tools_2.12-24.08.1.jar',
+                  'gs://bucket-name/path/to/file.jar'])
+    dependency_type: RuntimeDependencyType = Field(
+        default_factory=lambda: RuntimeDependencyType(dep_type=DependencyType.JAR),
```
Good catch! I was looking for this and I could not find it :)

parthosa reviewed (approved):
Thanks @amahussein. LGTME. The configuration module could be very useful in future. Setting runtime properties (e.g. thresholds for TCV, distributed spark configs etc)

nartal1 reviewed (dismissed):
Thanks @amahussein ! LGTM.

cindyyuanjiang commented on user_tools/src/spark_rapids_pytools/rapids/rapids_tool.py:
```diff
@@ -532,47 +544,46 @@ def exception_handler(future):
             return dep_item
 
-        def cache_all_dependencies(dep_arr: List[dict]):
+        def cache_all_dependencies(dep_arr: List[RuntimeDependency]):
```
nit: function return type

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/enums.py:
```diff
@@ -101,6 +101,21 @@ def get_hash_func(self) -> Callable:
+    CLASSPATH = 'classpath'
+
+    @classmethod
+    def get_default(cls):
```
nit: function return type

cindyyuanjiang reviewed (approved):
Thanks @amahussein! Very minor nits.

tgravescs commented:
one general comment about the description is that it doesn't seem like the section "The specification of the file is as follows".. matches what the real spec is.  Or atl east its very hard to read and match them up.  was this auto generated?

For instance:
```
DependencyType:
    description: Represents the dependency type for the jar cmd
    enum:
    - jar
    - archive
    - classpath
    title: DependencyType
    type: string
```
Description is dependency type for the jar command (what is jar command here?)... but enum is jar, archive, or classpath.  It doesn't seem to match dependency_type field in the example above it.

```
dependency_type:
        dep_type: archive
        relative_path: jars/*
```

tgravescs commented on user_tools/src/spark_rapids_tools/configuration/tools_config.py:
```diff
@@ -0,0 +1,63 @@
+                    'This is used to test the compatibility of the '
+                    'configuration file against the current tools release',
+        examples=['1.0'],
+        le=1.0,  # minimum version compatible with the current tools implementation
```
does this automatically fail to read of its not 1.0?  does it fail the entire tools run?

amahussein replied:
Yes, it will fail early in the phase in which we verify the arguments. As we move fwd, we can have a minimum compatible version. For example, backward compatibe with older formats can be expressed as `ge(1.1) and le(2.0)`
The point here is that the execution should not proceed because if the user is using out-daded configuration that does not fit the tools assumptions any more.

tgravescs commented:
when I run with a tools config file with only the info and api version it fails with:

```
Traceback (most recent call last):
....
TypeError: argument 'message_template': 'ValidationError' object cannot be converted to 'PyString'
```

Is there a way to get a more informative error message?

tgravescs commented on user_tools/src/spark_rapids_tools/cmdli/argprocessor.py:
```diff
@@ -370,6 +372,23 @@ def process_jvm_args(self) -> None:
 
+    def process_tools_config(self) -> None:
+        """
+        Load the tools config file if it is provided. it creates a ToolsConfig object and sets it
```
nit "it" should be capitalized. perhaps rename to like load_tools_config or set_tools_config so its not confusing that this is not "processing the actual dependencies"

tgravescs commented on user_tools/src/spark_rapids_tools/configuration/common.py:
```diff
@@ -0,0 +1,87 @@
+        validation_alias=AliasChoices('dep_type', 'depType'))
+    relative_path: str = Field(
+        default=None,
+        description='The relative path of the dependency in the classpath. This is relevant for tar files',
```
"This is relevant for tar files".. does this means I can't use .gz files or just .jar files?

amahussein replied:
I will clarify the description. This field is used to tell the tools which subdirectory to find the additional jars. Therefore, it is only relevant with "archive" like spark.tgz because it produces an entire directory.
if the dependency is `.jar` file , then such information is not needed because the jar file is added to the classpath.

amahussein commented on user_tools/src/spark_rapids_tools/configuration/common.py:
```diff
@@ -0,0 +1,87 @@
+
+
+class RuntimeDependency(BaseModel):
+    """The runtime dependency required by the tools Jar cmd. All elements are downloaded and added
```
Good idea! I am relacing the jar by Java

amahussein commented on user_tools/src/spark_rapids_tools/configuration/tools_config.py:
```diff
@@ -0,0 +1,63 @@
+
+class ToolsConfig(BaseModel):
+    """Main container for the user's defined tools configuration"""
+    info: ToolsConfigInfo = Field(
```
You are right. standard is the api-version be at the root level.

amahussein commented on user_tools/src/spark_rapids_tools/cmdli/tools_cli.py:
```diff
@@ -79,6 +80,7 @@ def qualification(self,
         :param jvm_threads: Number of threads to use for parallel processing on the eventlogs batch.
                 Default is calculated as a function of the total number of cores and the heap size on the host.
         :param verbose: True or False to enable verbosity of the script.
+        :param tools_config_file: Path to a configuration file that contains the tools' options.
```
Good Point. It is tricky to do it if we do not simply hardcode the file specification into the doc.
Let me think of how to hook that dynamically into the help output.

amahussein commented:
> one general comment about the description is that it doesn't seem like the section "The specification of the file is as follows".. matches what the real spec is. Or atl east its very hard to read and match them up. was this auto generated?
> 
> For instance:
> 
> ```
> DependencyType:
>     description: Represents the dependency type for the jar cmd
>     enum:
>     - jar
>     - archive
>     - classpath
>     title: DependencyType
>     type: string
> ```
> 
> Description is dependency type for the jar command (what is jar command here?)... but enum is jar, archive, or classpath. It doesn't seem to match dependency_type field in the example above it.
> 
> ```
> dependency_type:
>         dep_type: archive
>         relative_path: jars/*
> ```
> 
> I assume some fields are optional.. like dependency_type: is only listed in your example for one of the 3 dependencies?
> 
> Also the description doesn't detail how this interacts with the default spark downloads, please add those details.

Thanks @tgravescs 

I think the issue here that the autogenerated file picks both the field description and the class pydoc which causes it to be confusing.
There is some control on the autogenerated file by excluding fields, adding examples...etc. I will revisit it to pick a better explanation of the schema

amahussein commented:
Thanks @tgravescs 
As we discussed offline:

- I updated the description of fields
- I added unit tests to have clear understanding of valid/invalid files and the expected behavior
- I fixed the errors to be more clear. Specifying which fields are failing.
- I put a sample specification file in the test folders to be referenced. In addition, I added github link to the valid yaml files in the output of the help cmd.

New specifications after the changes for convenience:
https://github.com/NVIDIA/spark-rapids-tools/pull/1395/files#diff-79a9ce702cf001815401a8737b8ec035c1f7ed01d73a5360f2b4f0142807be5b

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/cmdli/argprocessor.py:
```diff
@@ -384,10 +384,13 @@ def process_tools_config(self) -> None:
+                    'invalid_config',
                     f'Tools config file path {self.tools_config_path} could not be loaded. '
-                    'It is expected to be a valid YAML file.\n  Error:') from ve
+                    f'It is expected to be a valid configuration YAML file.'
```
nit: `f'It is expected to be a valid configuration YAML file.'` does not need formatting.

amahussein replied:
Fixed!

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/configuration/common.py:
```diff
@@ -23,33 +23,28 @@
+                    'Requires field dep_type to be set to (archive).',
         validation_alias=AliasChoices('relative_path', 'relativePath'),
-        examples=['jars/*'])
+        examples=['*'])
```
Q: I am not sure if this helps me understand what `relative_path` should be?

amahussein replied:
Fixed putting back `jars/*`

There is description on `dependency_type`.

> "Specifies the dependency type to determine how the item is processed. For example, jar files are appended to the java classpath while archive files such as spark are extracted first before adding subdirectory _/jars/* to the classpath"

Do you have a suggestion on a better way to document that? The problem is that adding too much details make the file hard to read.

cindyyuanjiang replied:
thanks @amahussein! this looks better to me.

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/configuration/common.py:
```diff
@@ -59,29 +54,21 @@ class DependencyVerification(BaseModel):
     verification: DependencyVerification = Field(
         default=None,
-        description='The verification information of the dependency.',
-        examples=[
```
nit: are the examples removed on purpose?

amahussein replied:
Yes, because the examples were not in the correct level.
This caused some confusion in understanding the file.

cindyyuanjiang replied:
thanks @amahussein!

cindyyuanjiang reviewed (commented):
Thanks @amahussein! Some minor nits and questions.

parthosa reviewed (approved):
Thanks @amahussein. This was a nice approach.

cindyyuanjiang reviewed (approved):
Thanks @amahussein for adding this feature!
